### PR TITLE
Remove all references to optionalt

### PR DESCRIPTION
**Pull request by tautschnig**
We now use `std::optional` instead.

I am not sure it's worth the code churn to review/merge this at this point. Most of the changes were implemented using `git grep -wl optionalt | xargs sed -i 's/optionalt/std::optional/g'` followed by removal of `#include <util/optional.h>` and then manually re-adding `#include <optional>` as needed (plus clang-format). My main reason for even embarking on this was to figure out what changes are needed so as to get rid of `-Wno-deprecated-declarations`.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/jar_file.h
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/jar_file.h:16]: ⛏️ Shouldn't this be replaced by `#include <optional>`? If this is working at the moment, I'd guess it was down to transitive includes.

File: src/goto-harness/memory_snapshot_harness_generator.h
Comment (thomasspriggs) [inline src/goto-harness/memory_snapshot_harness_generator.h:12]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-harness/recursive_initialization.h
Comment (thomasspriggs) [inline src/goto-harness/recursive_initialization.h:17]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-instrument/contracts/dynamic-frames/dfcc.cpp
Comment (thomasspriggs) [inline src/goto-instrument/contracts/dynamic-frames/dfcc.cpp:19]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.h
Comment (thomasspriggs) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.h:20]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_library.h
Comment (thomasspriggs) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_library.h:17]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_wrapper_program.h
Comment (thomasspriggs) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_wrapper_program.h:20]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-instrument/cover_basic_blocks.h
Comment (thomasspriggs) [inline src/goto-instrument/cover_basic_blocks.h:15]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-instrument/unwindset.h
Comment (thomasspriggs) [inline src/goto-instrument/unwindset.h:16]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-programs/format_strings.h
Comment (thomasspriggs) [inline src/goto-programs/format_strings.h:20]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-programs/remove_virtual_functions.h
Comment (thomasspriggs) [inline src/goto-programs/remove_virtual_functions.h:18]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-programs/resolve_inherited_component.h
Comment (thomasspriggs) [inline src/goto-programs/resolve_inherited_component.h:17]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-programs/restrict_function_pointers.h
Comment (thomasspriggs) [inline src/goto-programs/restrict_function_pointers.h:25]: ⛏️ See previous comment about `#include <optional>`

File: src/goto-programs/show_properties.h
Comment (thomasspriggs) [inline src/goto-programs/show_properties.h:16]: ⛏️ See previous comment about `#include <optional>`

File: src/pointer-analysis/add_failed_symbols.h
Comment (thomasspriggs) [inline src/pointer-analysis/add_failed_symbols.h:16]: ⛏️ See previous comment about `#include <optional>`

File: src/util/arith_tools.h
Comment (thomasspriggs) [inline src/util/arith_tools.h:15]: ⛏️ See previous comment about `#include <optional>`

File: src/util/dense_integer_map.h
Comment (thomasspriggs) [inline src/util/dense_integer_map.h:20]: ⛏️ See previous comment about `#include <optional>`

**Review by thomasspriggs — APPROVED** [review]
Getting this merged for version 6 should be given a reasonably high priority. The reason being that you cannot construct or assign an `optionalt` using an `std::optional` (or the other way around.) So if we don't get rid of `optionalt`, then this could lead to awkward maintenance where maintainers need to consider which type of optional to use in which place.

File: jbmc/src/java_bytecode/jar_file.h
Comment (tautschnig) [reply jbmc/src/java_bytecode/jar_file.h:16]: It seemed to work fine on my Ubuntu 20.04, but other OS/compiler combinations indeed make those includes necessary. I had naively assumed that standard library headers would likely (transitively) include that anyway, but as experimenting shows that's not always the case. So, yes, I may well end up putting this back in.

File: src/goto-programs/format_strings.h
Comment (tautschnig) [reply src/goto-programs/format_strings.h:20]: Done.

File: src/goto-programs/resolve_inherited_component.h
Comment (tautschnig) [reply src/goto-programs/resolve_inherited_component.h:17]: Done.

**Comment by kroening** [comment]
> I am not sure it's worth the code churn

Maybe not under normal circumstances, but I'd be inclined to say that the transition to 6.0 makes it appropriate.

File: src/goto-programs/show_properties.h
Comment (tautschnig) [reply src/goto-programs/show_properties.h:16]: Done.

File: jbmc/src/java_bytecode/jar_file.h
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/jar_file.h:16]: Good to see this being fixed.

I'd prefer to avoid relying on transitive header includes where possible. With `optional` as an example my reasoning goes something like -

1. Header "A" depends on `optional` so `#include <optional>` is added to it.
2. `.cpp` file "B" depends on header "A" and uses `std::optional` in its implementation. It therefore depends on `optional`, but as this `.cpp` already includes header "A" `#include <optional>` is not added to it.
3. Maintainer of header "A" decides they no longer need to depend on `optional`. Lets say they were using it for the type of a private member function rather than as part of the exposed API. They remove the `#include <optional>` from header "A", because it no longer needs it itself.
4. The compilation of `.cpp` "B" now fails. This means that the maintainer of header "A" now needs to either add the include to the `.cpp` file, or to leave the extra include in their header.

Leaving the extra include in could potentially extend compile times. Adding it to the `.cpp` increases the number of changes in their PR and potentially requires approval from more code owners. The decision and additional work could have been avoided if the reliance on transitive includes had been avoided.

File: src/goto-instrument/unwindset.h
Comment (tautschnig) [reply src/goto-instrument/unwindset.h:16]: Done.

**Review by peterschrammel — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]
LGTM in all contracts related changes.

**Review by kroening — APPROVED** [review]

**Pull request merged**
